autosubmit/statistics/jobs_stat.py

```diff
@@ -1,18 +1,65 @@
-#!/bin/env/python
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
 from datetime import datetime, timedelta
-from .utils import timedelta2hours
+from math import ceil
+from typing import Optional
+
+from autosubmit.statistics.utils import timedelta2hours
 from log.log import Log
-import math
+
+
+def _estimate_requested_nodes(nodes, processors, tasks, processors_per_node) -> int:
+    if str(nodes).isdigit():
+        return int(nodes)
+    elif str(tasks).isdigit():
+        return ceil(int(processors) / int(tasks))
+    elif str(processors_per_node).isdigit() and int(processors) > int(processors_per_node):
+        return ceil(int(processors) / int(processors_per_node))
+    else:
+        return 1
+
+
+def _calculate_processing_elements(nodes, processors, tasks, processors_per_node, exclusive) -> int:
+    if str(processors_per_node).isdigit():
+        if str(nodes).isdigit():
+            return int(nodes) * int(processors_per_node)
+        else:
+            estimated_nodes = _estimate_requested_nodes(nodes,processors,tasks,processors_per_node)
+            if not exclusive and estimated_nodes <= 1 and int(processors) <= int(processors_per_node):
+                return int(processors)
+            else:
+                return estimated_nodes * int(processors_per_node)
+    elif str(tasks).isdigit() or str(nodes).isdigit():
+        Log.warning(f'Missing PROCESSORS_PER_NODE. Should be set if TASKS or NODES are defined. '
+                    f'The PROCESSORS will used instead.')
+    return int(processors)
```

Comment:
These functions were in the class but had zero dependency on the class or its state. Instead of making them static and part of an class/object, we move it to the module which makes it a lot simpler to write unit tests for, and also gives us some small/tiny/really small performance improvement, I believe.